### DISCUSSION TITLE
How much do I need the BitTorrent port and how do I change it?

### DISCUSSION THREAD

**Discussion by ioqy** [bitmagnet-io/bitmagnet#148] [Q&A] [answered]
In the docker file a "BitTorrent port" 3334 is exposed but nowhere is explained if this port is needed for bitmagnet to work or not and how to change it.

Since I don't have the port exposed and I'm finding torrents, I'm suspecting that the port is not needed for bitmagnet to work, but with it, bitmagnet might work more efficiently. Is that correct?

In the [full docker compose example](https://github.com/bitmagnet-io/bitmagnet/blob/main/docker-compose.yml) the port 3334 is also exposed on the gluetun container. This doesn't make sense to me, because it would open the container on the host and therefore bypass the VPN connection of gluetun. I guess this is an error?

Is there any way to change the port? I did a quick search through the code but couldn't find anything that would indicate this.
I know that the port can be changed by changing the left exposed port in the docker compose file, but only when not using gluetun. Since I'm using gluetun this is not an option.

**Comment by kzshantonu** [comment]
I'd like to know about this too, if the port needs to be publicly accessible, there should be a way to set a custom port number

**Comment by mgdigital** [comment, marked as answer]
> Since I don't have the port exposed and I'm finding torrents, I'm suspecting that the port is not needed for bitmagnet to work, but with it, bitmagnet might work more efficiently. Is that correct?

I'm aware that Bitmagnet works without this port exposed for UDP in Docker. To be honest I'm not really clear on why it works without being exposed (anyone else able to shed light on this?), but I have left it in the docker compose file as an indicator that it's being used, and in case not exposing it caused issues for anyone...

I've also added TCP there too, intended for servicing peer protocol requests. Nothing is implemented for this at the moment but I figured best to put it there from the start and avoid having to tell people to update their config later...

The port number is configurable with the `dht_server.port` setting.

**Reply by chkuendig** [reply]
Theres a few ways this can work, see https://en.m.wikipedia.org/wiki/UDP_hole_punching

If at least one of two peers has an open port its even easier as that peer can do the "hole punching coordination".

I yesterday worked on bitmagnet on the train over LTE with gnat (and i guess my phone doing yhe tethering adding another layer) and it worked just fine :)

**Reply by chkuendig** [reply]
Small Postscript: The details on how this works for the DHT specifically are defined in BEP55: http://bittorrent.org/beps/bep_0055.html

**Reply by mgdigital** [reply]
However - firstly, Bitmagnet has not implemented BEP55. I know a little about it - especially that it's very non-trivial to implement. Secondly it is a peer protocol thing (i.e. TCP, not UDP).

So I'm still a bit confused why this works. Is the UDP hole punching linked from Wikipedia something that happens automatically at the network level? My other theory was that it's something to do with routing to your VPN via Gluetun...

**Reply by Send8213** [reply]
My understanding is that since UDP is a connectionless-protocol, the packets can't be linked to a specific existing established connection in firewalls/NAT/etc like they can with TCP. Most firewalls only block incoming stuff but allow outgoing connections and incoming data related to those outgoing connections to come back in. To work around this in order to not break one of the protocols the internet runs on, a firewall for UDP will keep track of (src ip, src port, dst ip, dst port) combinations and allow traffic if there has been some kind of outgoing request that matches. If both sides send an outgoing request to each other at the same time, either coordinated by some third-party or just randomly, then to each of the firewalls it's just outgoing traffic and associated returning traffic and is therefore allowed.

As far as the protocol is concerned, hole punching probably only works with the uTP (BEP29) over UDP and DHT, not the traditional TCP bittorrent protocol.

I believe the networking stack under docker does this kind of UDP established connection tracking like this and that's what lets packets in and why it can still work.
